common/ferrostar/src/algorithms.rs

```diff
@@ -256,6 +259,60 @@ pub fn distance_to_end_of_step(
     }
 }
 
+/// Computes the arrival state for a snapped location along the route.
+/// This includes distances and durations.
+pub fn calculate_arrival(
+    snapped_location: &Point,
+    remaining_steps: &[RouteStep],
+) -> ArrivalState {
+    if remaining_steps.is_empty() {
+        return ArrivalState {
+            distance_to_next_maneuver: 0.0,
+            distance_remaining: 0.0,
+            duration_remaining: 0.0,
+        };
+    }
+
+    let mut distance_remaining = 0.0;
+    let mut duration_remaining = 0.0;
+
+    let current_step = remaining_steps.first().unwrap();
+    let distance_to_next_maneuver = distance_to_end_of_step(snapped_location, &current_step.get_linestring());
+    // Add the current step's distance to the total distance remaining
+    distance_remaining += distance_to_next_maneuver;
+
+    // TODO: Improve this logic? It may be solid, but seems we could improve it with actual speed data.
+    let pct_remaining_current_step = distance_to_next_maneuver / current_step.get_linestring().haversine_length();
+
+    // Get the percentage of duration remaining in the current step.
+    let duration_to_end_of_step = pct_remaining_current_step * current_step.duration;
+    // Add the current step's duration to the total distance remaining
+    duration_remaining += duration_to_end_of_step;
+    
+    // Exit early if there is only the current step:
+    if remaining_steps.len() == 1 {
+        return ArrivalState {
+            distance_to_next_maneuver,
+            distance_remaining,
+            duration_remaining
+        };
+    }
+
+    // Loop through the steps after the current and add their total duration/distance
+    let steps_after_current = &remaining_steps[1..];
+    for step in steps_after_current {
+        distance_remaining += step.distance;
+        duration_remaining += step.duration;
+    }
+
+    ArrivalState {
+        distance_to_next_maneuver,
+        distance_remaining,
+        duration_remaining
+    }
```

Comment:
I think this can be effectively be replaced with a reduction rather than using mutable variables.

I can probably take a swing at it before the meeting.